lib/trino-filesystem/src/main/java/io/trino/filesystem/UnrecoverableIOException.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.filesystem;
+
+import com.google.common.base.Throwables;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Optional;
+
+/**
+ * This exception is for stopping retries for FileSystem calls that shouldn't be retried.
+
+ * Use this exception for cases, where explicit message that IO has failed and retrying doesn't make any sense in context of operation success.
+ * There could be various reasons such as 404, 403, 401 related error codes.
+
+ * By default, all third party (AWS, Azure, GCP) SDKs will retry appropriate exceptions (either client side IO errors, or 500/503), so there is
+ * no need to retry those additionally.
+ * If any custom retry behavior is needed, it is advised to change SDK's retry handlers, rather than introducing outer retry loop, which combined
+ * with SDKs default retry, could lead to prolonged, unnecessary retries
+ */
+public class UnrecoverableIOException
```

Comment:
I'd call this `TrinoFilesystemException`. It's not necessarily IO specific (403 or 404 errors are not IO), and there's the `isUnrecoverableIOException` method.

Reply:
403 or 404 errors are IO in the sense that they are doing IO to a remote system. It is a subclass of `IOException` as we use that checked exception to ensure these excepted errors are handled.

Given that there are other unrecoverable errors, specifically `FileNotFoundException`, naming it `TrinoFileSystemException` seems like a good idea. Naming it `Unrecoverable` could imply that it covers all unrecoverable exceptions. Instead, the utility method needs to be used.

Reply:
Let's add the `isUnrecoverableException` method to `TrinoFileSystem` in this commit.